gateway/api/connections/connections.go

```diff
@@ -638,45 +636,203 @@ func GetDatabaseSchemas(c *gin.Context) {
-			}
-			c.Writer.Header().Set("Content-Length", fmt.Sprintf("%v", contentLength))
+	// Validate database name to prevent SQL injection
+	if err := validateDatabaseName(dbName); err != nil {
+		c.JSON(http.StatusUnprocessableEntity, gin.H{"message": err.Error()})
+		return
+	}
+
+	// If the schema is not provided, use 'public' for PostgreSQL and the same database for other databases
+	if schemaName == "" {
+		schemaName = dbName
+		if currentConnectionType == pb.ConnectionTypePostgres {
+			schemaName = "public"
+		}
+	}
+
+	script := getColumnsQuery(currentConnectionType, dbName, tableName, schemaName)
+	if script == "" {
+		c.JSON(http.StatusBadRequest, gin.H{"message": "unsupported database type"})
+		return
+	}
+
+	userAgent := apiutils.NormalizeUserAgent(c.Request.Header.Values)
+	client, err := clientexec.New(&clientexec.Options{
+		OrgID:          ctx.GetOrgID(),
+		ConnectionName: conn.Name,
+		BearerToken:    getAccessToken(c),
+		UserAgent:      userAgent,
+		Verb:           pb.ClientVerbPlainExec,
+	})
+	if err != nil {
+		log.Error(err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
+		return
+	}
+
+	respCh := make(chan *clientexec.Response)
+	go func() {
+		defer func() { close(respCh); client.Close() }()
+		select {
+		case respCh <- client.Run([]byte(script), nil):
+		default:
+		}
+	}()
+
+	timeoutCtx, cancelFn := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancelFn()
+	select {
+	case outcome := <-respCh:
+		if outcome.ExitCode != 0 {
+			log.Warnf("failed issuing plain exec: %s, output=%v", outcome.String(), outcome.Output)
+			c.JSON(http.StatusBadRequest, gin.H{"message": fmt.Sprintf("failed to get columns: %s", outcome.Output)})
 			return
 		}
-		c.JSON(http.StatusOK, schema)
+
+		response := openapi.ColumnsResponse{Columns: []openapi.ConnectionColumn{}}
+
+		if currentConnectionType == pb.ConnectionTypeMongoDB {
+			// Parse MongoDB output
+			columns, err := parseMongoDBColumns(outcome.Output)
+			if err != nil {
+				log.Errorf("failed parsing mongo response: %v", err)
+				c.JSON(http.StatusInternalServerError, gin.H{"message": fmt.Sprintf("failed to parse MongoDB response: %v", err)})
+				return
+			}
+			response.Columns = columns
+		} else {
+			// Parse SQL output
+			columns, err := parseSQLColumns(outcome.Output, currentConnectionType)
+			if err != nil {
+				log.Errorf("failed parsing SQL response: %v", err)
+				c.JSON(http.StatusInternalServerError, gin.H{"message": fmt.Sprintf("failed to parse SQL response: %v", err)})
+				return
+			}
+			response.Columns = columns
+		}
+
+		c.JSON(http.StatusOK, response)
 
 	case <-timeoutCtx.Done():
 		client.Close()
-		log.Infof("runexec timeout (50s), it will return async")
-		c.JSON(http.StatusBadRequest, gin.H{"message": "Request timed out"})
+		log.Warnf("timeout (30s) obtaining columns for table '%s' in database '%s' using connection '%s'", tableName, dbName, conn.Name)
+		c.JSON(http.StatusRequestTimeout, gin.H{
```

Comment:
This status code should be used when the "server didn't receive a request from the client within the expected timeframe". This is not the case here. Change to a 400 status code instead